### PR TITLE
CON-1122: Edit the security doc

### PR DESCRIPTION
**Pull request by JohnJoser3**
Edit the security doc as per the findings in the content audit.

### CODE REVIEW

File: docs/docs/security.md [hunk 1]
```diff
@@ -1,167 +1,162 @@
...
+previously presented time. If the host rewinds the clock, the enclave time will pause until the host time again
+catches up with the latest time in the enclave. However, if the enclave restarts, the time inside the enclave
+could be earlier than in a previous instance. Additionally, the host may pause the execution of
+an enclave at any moment for an arbitrary duration without the enclave code being aware of that. So, if you were to
+read a timestamp, it might be very old by the time you use it.
 
-However the protocol for accessing the timestamp doesn't involve doing any calls out of the enclave, so, the host can't
-observe the program's progress by watching OCALLs (which would otherwise be a side channel).
+However, the protocol for accessing the timestamp doesn't involve doing any calls out of the enclave. So, the host can't
+observe the program's progress by watching Outside Calls (OCall).
```
Comment (shamsasari) [inline docs/docs/security.md:24]: Have the mention of "side channel"

Comment (shamsasari) [inline docs/docs/security.md:24]: "OCALL"

File: docs/docs/security.md [hunk 2]
```diff
@@ -1,167 +1,162 @@
...
+
+However, there's one class of attack encryption cannot stop. This attack is when the host gives you back older data than
+was requested. The owner of the computer controls the system clock, which makes it unreliable. Additionally,
+the owner can kill and restart the host process whenever they like. This means that a system owner can tamper with 
+the enclave's sense of time and order of events. By snapshotting the stored (sealed, encrypted) data an enclave has 
+provided after each network message from a client is delivered, the enclave can be "rewound" to any
+point. Then, stored messages from the clients can be replayed back to the enclave in different orders, or with some
+messages dropped.
 
+Such attacks are called memento/rewind/replay attacks. You need to consider this type of attack while designing your
```
Comment (shamsasari) [inline docs/docs/security.md:52]: Update this to mention the existing feature we have to mitigate this attack and link to relevant section in persistence.md.

File: docs/docs/security.md [hunk 3]
```diff
@@ -1,167 +1,162 @@
...
-* **Storage access patterns**. If your enclave doesn't access the database when processing a message of type A, but does
-  when processing a message of type B, or accesses the database with a different sequence, number or type of accesses,
-  the host can learn the message type by observing those accesses.  
+* **Message sizes**. If an enclave is known to process only two kinds of message of around 1 kilobyte or 100 
+  kilobytes, then the size of the encrypted message by itself leaks information about what kind of message it is.
+* **Message processing time**. If an enclave processes a message type in 1 millisecond and another message type in 
+  100 milliseconds, observing how much work the enclave does for a new message, can reveal what kind of message it is.
+* **Storage access patterns**. Suppose an enclave doesn't access the database when processing a message of type A, 
+  but does when processing a message of type B, or accesses the database with a different sequence, number or type 
+  of accesses. In that case, the host can learn the message type by observing those accesses.
```
Comment (shamsasari) [inline docs/docs/security.md:103]: Add that Conclave randomises the sector accesses when the enclave uses the persistence file system to reduce this side channel for the user. It's mentioned in persistence.md.

File: docs/docs/security.md [hunk 4]
```diff
@@ -1,167 +1,162 @@
...
+Malicious actors can use these techniques to reveal fine-grained information as well. For instance, if an encrypted
+piece of data contains a number used to control a loop that does data lookups, counting the number of external data
+lookups reveals the number.
 
-Some kinds of architectural side channel attacks can be mitigated by Conclave for you, for instance, messages can be
-padded so all encrypted messages look the same size. This sort of approach works if message sizes don't vary too wildly
-and you can afford to use the bandwidth and storage to set all messages to their maximum possible size. How valuable it
-is in your situation is a topic for you to consider as you design your app.
+Conclave can mitigate some of these architectural side-channel attacks. For instance, Conclave can pad all messages 
+to make all encrypted messages appear the same size. This approach works if message sizes don't vary wildly and you
```
Comment (shamsasari) [inline docs/docs/security.md:110]: Would it make more sense to mention this mitigation that Conclave provides in the `Message sizes` bullet point above?

File: docs/docs/security.md [hunk 5]
```diff
@@ -1,167 +1,162 @@
...
-mechanisms and leak data out of the enclave - not via normal means (which the CPU doesn't allow) but by affecting the
-timing of subsequent operations which can then be measured.
+There are a variety of different attacks with varying details. These attacks work by exploiting the speculative
+execution capability of the processor to make an enclave compute on invalid data, which can then be used to bypass 
+security mechanisms and leak data out of the enclave - not via normal means (which the CPU doesn't allow) but
+by affecting the timing of subsequent operations, which can then be measured.
 
-Micro-architectural side channel attacks can be resolved at a layer lower than the architecture of your own application.
-They often require reducing the performance of either the enclave or the entire host system however, so it's worth
+You can prevent Micro-architectural side-channel attacks at a layer lower than the architecture of your application.
```
Comment (shamsasari) [inline docs/docs/security.md:124]: Micro -> micro

File: docs/docs/security.md [hunk 6]
```diff
@@ -1,167 +1,162 @@
...
-class XORs the system time with a truly random number derived from the `RDRAND` instruction. This results in
-the class using a truly unique high entropy seed. Subsequent instances of `Random` transform the seed to
-ensure no two sequences are the same but the entropy is not renewed using the `RDRAND` instruction.
-Therefore for the best quality random data it is still recommended to use `SecureRandom`.
\ No newline at end of file
+To generate random numbers inside the enclave, you need to use a strong random number generator that cannot be
+weakened by the host.
+
+The safest way is to use the `SecureRandom` class to generate random numbers. In Conclave enclaves, the 
+implementation of this class uses the `RDRAND` instruction to use an on-CPU hardware random number generator.
```
Comment (shamsasari) [inline docs/docs/security.md:152]: to use -> which uses

File: docs/docs/security.md [hunk 7]
```diff
@@ -1,167 +1,162 @@
...
-an insecure random number generator, Conclave modifies this behaviour so the first instance of a `Random`
-class XORs the system time with a truly random number derived from the `RDRAND` instruction. This results in
-the class using a truly unique high entropy seed. Subsequent instances of `Random` transform the seed to
-ensure no two sequences are the same but the entropy is not renewed using the `RDRAND` instruction.
-Therefore for the best quality random data it is still recommended to use `SecureRandom`.
\ No newline at end of file
+To generate random numbers inside the enclave, you need to use a strong random number generator that cannot be
+weakened by the host.
+
+The safest way is to use the `SecureRandom` class to generate random numbers. In Conclave enclaves, the
```
Comment (shamsasari) [inline docs/docs/security.md:151]: Link the SecureRandom docs: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/security/SecureRandom.html

**Review by shamsasari — CHANGES_REQUESTED** [review]

File: docs/docs/security.md (hunk 1 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: Added.

Comment (JohnJoser3) [reply docs/docs/security.md]: Updated.

File: docs/docs/security.md (hunk 2 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: Updated.

File: docs/docs/security.md (hunk 3 above)
Comment (JohnJoser3) [reply docs/docs/security.md:102]: Added.

File: docs/docs/security.md (hunk 4 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: Done.

File: docs/docs/security.md (hunk 6 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: done.

File: docs/docs/security.md (hunk 7 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: Done.

File: docs/docs/security.md [hunk 8]
```diff
@@ -49,14 +49,8 @@ provided after each network message from a client is delivered, the enclave can
...
-Such attacks are called memento/rewind/replay attacks. You need to consider this type of attack while designing your
-app.
-
-!!!Warning
-
-    A full discussion of memento attacks is beyond the scope of this document. The Conclave project strives to provide
-    you with high-level protocol constructs that take them into account, but when writing your own protocols, you should
-    consider what happens when messages can be re-ordered and re-tried by the host.
+Such attacks are called memento/rewind/replay attacks.
+You can use Conclave's [persistent map](persistence.md#PersistentMap) to mitigate rewind attacks.
```
Comment (shamsasari) [inline docs/docs/security.md:53]: The link isn't correct. I believe the heading link should be `#persistent-map`

File: docs/docs/security.md [hunk 9]
```diff
@@ -96,21 +90,23 @@ Here is a non-exhaustive set of examples:
...
+  if message sizes don't vary wildly, and you can afford to use the bandwidth and storage required to set all messages
+  to their maximum possible size.
+
 * **Message processing time**. If an enclave processes a message type in 1 millisecond and another message type in 
   100 milliseconds, observing how much work the enclave does for a new message, can reveal what kind of message it is.
+
 * **Storage access patterns**. Suppose an enclave doesn't access the database when processing a message of type A, 
   but does when processing a message of type B, or accesses the database with a different sequence, number or type 
   of accesses. In that case, the host can learn the message type by observing those accesses.
+  Conclave's [persistent file system](persistence.md#PersistentEncryptedFilesystem) randomizes sector accesses to
```
Comment (shamsasari) [inline docs/docs/security.md:103]: ditto. I believe this should be `#persistent-encrypted-filesystem`.

**Review by shamsasari — CHANGES_REQUESTED** [review]

File: docs/docs/security.md (hunk 9 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: Corrected.

File: docs/docs/security.md (hunk 8 above)
Comment (JohnJoser3) [reply docs/docs/security.md]: Corrected.

**Review by shamsasari — APPROVED** [review]
Thanks

**Pull request merged**